storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -166,93 +257,182 @@ func (s *commitLogSource) Read(
 					lastWriteAt: dp.Timestamp,
 					enc:         enc,
 				})
-				unmergedSeries.encoders[blockStart] = unmergedBlock
+				unmergedSeries.encoders[blockStart.UnixNano()] = unmergedBlock
 			}
 		}
 		if err != nil {
-			errs++
+			workerErrs[workerNum]++
 		}
 	}
-	if errs > 0 {
-		s.log.Errorf("error bootstrapping from commit log: %d block encode errors", errs)
+	wg.Done()
+}
+
+func (s *commitLogSource) shouldEncodeSeries(
+	namespace ts.ID,
+	shardsTimeRanges result.ShardTimeRanges,
+	blockSize time.Duration,
+	series commitlog.Series,
+	dp ts.Datapoint,
+) bool {
+	// Check if the series belongs to current namespace being bootstrapped
+	if !namespace.Equal(series.Namespace) {
+		return false
+	}
+
+	// Check if the shard is one of the shards we're trying to bootstrap
+	ranges, ok := shardsTimeRanges[series.Shard]
```

Comment:
If we allocate unmerged upfront and only allocate maps for shards in that slice we can do a slice lookup here which is much faster than a map lookup.

Reply:
nice I like it.

Reply:
ah actually, I still need access to the shard's ranges to perform the range overlap check at the end. In that case, it actually becomes slower because we do a slice lookup followed by a map lookup. I could probably copy the ranges into the unmerged slice though so that they're both available for quick lookup. Might be worth it since this is on the hot path

Reply:
ended up creating the encodersAndRanges struct and doing it that way